src/http/ngx_http_request.c

```diff
@@ -2799,7 +2799,7 @@ ngx_http_finalize_connection(ngx_http_request_t *r)
     }
 
     if (!ngx_terminate
-         && !ngx_exiting
+         && (!ngx_exiting || clcf->keepalive_during_shutdown)
```

Comment:
This might be unnecessary due to https://github.com/nginx/nginx/blob/29aec5720fdfc74dca8d99d5cf6dc0fcb4e4ce2f/src/http/ngx_http_header_filter_module.c#L201

Reply:
Nevermind, the exit request can come after headers processing.

Reply:
Exactly